server/api/social/link.js

```diff
@@ -0,0 +1,114 @@
+import { Router } from 'express';
+import { jwtAuth } from '../../util/jwt';
+import { ValidationError } from '../../../util/ValidationHelper';
+
+const { userCollection: dbRef } = require('../../util/firebase');
+
+const router = Router();
+
+router.post('/social/links/new', jwtAuth, async (req, res, next) => {
+  try {
+    const {
+      user,
+      link,
+    } = req.body;
+    if (req.user.user !== user) {
+      res.status(401).send('LOGIN_NEEDED');
+      return;
+    }
+
+    const {
+      id,
+      siteDisplayName,
+      order,
+      url,
+    } = link;
+    const isExternalLink = true;
+    const isPublic = true; // default
+
+    const linkDoc = await dbRef.doc(user).collection('social').doc(id).get();
+    if (linkDoc.exists) throw new ValidationError('link already exists');
+
+    await dbRef.doc(user).collection('social').doc(id).create({
+      isExternalLink,
+      isPublic,
+      order,
+      siteDisplayName,
+      url,
+    });
+
+    res.json({
+      id,
+      isExternalLink,
+      isPublic,
+      order,
+      siteDisplayName,
+      url,
+    });
+  } catch (err) {
+    next(err);
+  }
+});
+
+router.put('/social/links/:linkId', jwtAuth, async (req, res, next) => {
+  try {
+    const {
+      user,
+      link,
+    } = req.body;
+    if (req.user.user !== user) {
+      res.status(401).send('LOGIN_NEEDED');
+      return;
+    }
+
+    const { linkId } = req.params;
+    const {
+      siteDisplayName,
+      url,
+      order,
+    } = link;
+
+    const linkDoc = await dbRef.doc(user).collection('social').doc(linkId).get();
+    if (!linkDoc.exists) throw new ValidationError('link unknown');
+
+    const promises = [];
+
+    const updateObj = {};
+    if (siteDisplayName) updateObj.siteDisplayName = siteDisplayName;
+    if (url) updateObj.url = url;
+    if (order) {
+      updateObj.order = order;
+
+      const oldOrder = linkDoc.data().order;
+      let query;
+      let orderOffset;
+      if (oldOrder > order) {
+        query = await dbRef.doc(user).collection('social')
+          .where('order', '>=', order)
+          .where('order', '<', oldOrder)
+          .get();
+        orderOffset = 1;
+      } else if (oldOrder < order) {
+        query = await dbRef.doc(user).collection('social')
+          .where('order', '>', oldOrder)
+          .where('order', '<=', order)
+          .get();
+        orderOffset = -1;
```

Comment:
@nnkken do we have a better way of doing this...

Reply:
@mileswong even if we have to handle this `order` thingy like this
this have to be in a firestore transaction, otherwise it can race like hell
But I think we should figure out some better way to do ordering, instead adding it one by one...